server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/KMeans.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.codec.vectors.cluster;
+
+import org.apache.lucene.index.FloatVectorValues;
+import org.apache.lucene.util.VectorUtil;
+
+import java.io.IOException;
+import java.util.Random;
+
+/**
+ * k-means implementation specific to the needs of the {@link HierarchicalKMeans} algorithm
+ */
+class KMeans {
+
+    final int sampleSize;
+    final int maxIterations;
+
+    KMeans(int sampleSize, int maxIterations) {
+        this.sampleSize = sampleSize;
+        this.maxIterations = maxIterations;
+    }
+
+    /**
+     * uses a Reservoir Sampling approach to picking the initial centroids which are subsequently expected
+     * to be used by a clustering algorithm
+     *
+     * @param vectors used to pick an initial set of random centroids
+     * @param centroidCount the total number of centroids to pick
+     * @return randomly selected centroids that are the min of centroidCount and sampleSize
+     * @throws IOException is thrown if vectors is inaccessible
+     */
+    static float[][] pickInitialCentroids(FloatVectorValues vectors, int centroidCount) throws IOException {
+        Random random = new Random(42L);
+        int centroidsSize = Math.min(vectors.size(), centroidCount);
+        float[][] centroids = new float[centroidsSize][vectors.dimension()];
+        for (int i = 0; i < vectors.size(); i++) {
+            float[] vector = vectors.vectorValue(i);
```

Comment:
let's only read in the vector value if its used for sampling. Otherwise we are always read in every vector, which isn't strictly necessary.